### PR TITLE
gh-83151: Make closure work on pdb

### PR DESCRIPTION
**Pull request by gaogaotiantian**
Closure on pdb has been an issue for years, and is often considered "to difficult to fix". PEP709 alleviated the issue by inlining all list, dict and set comprehensions, but the fundamental issue is still there - you can easily reproduce the issue with lambda functions (`(lambda: x)()`) or generators (`any(x for x in lst if x != z)`). Now that `pdb` supports multi-line statement, you can even create your own function scope. Neither of them work in current `pdb`.

The related (open) issues I can find are #83151, #80052, #76987, #70260 and #65360.

This PR fixes all the issues mentioned above, while keep (almost?) all the current correct behaviors.

Boiler alert, black magic involved.

The fundamental reason for the issue above, is that when python compiles code that generates nested functions, the inner function has no idea about the context. For example:

```python
def f():
    z = 1
    any([x for x in range(2) if x != z])
```

The code above works fine because python compiles `any([x for x in range(2) if x != z])` with the awareness that `z` is a local variable of `f`, so closure should be used. However, in `pdb`, if you execute/compile `any([x for x in range(2) if x != z])`, there's no way for Python to know that `z` is an outer local variable and should be passed as a cell variable.

So, the solution is - trick the compiler.

What we can do, is to force compiler to compile the code we need in a closure-awareness environment. Take the code above as an example, instead of compiling `any([x for x in range(2) if x != z])`, we compile the following code:

```python
def outer():
    z = None
    def __pdb_scope():
        nonlocal z
        any([x for x in range(2) if x != z])
```

Then we get the code object of `outer.__pdb_scope`, and that is the code object we actually want - it considers `z` as a free variable. Then, we pass the value of `z` with `closure` argument of `exec`:

```python
exec(code, globals, locals, closure=(types.CellType(1),))
```

The solution is not done, as we still need to write the local variables back (`locals` won't be changed as all the variables are considered freevar). Also, it's possible for the code itself to create new variables and we want that in locals as well (`k=1` for example). So, we just write all the local variables in the scope back by adding some new code at the end of the code.

To keep the original behavior as close as possible, if any exception is raised in the new method, we fallback to the original simple `exec` solution (it might be a valid exception, but we will always have the same exception as before).

Why not ...
* just combine `globals()` and `locals()` as the global namespace?
    * result of `globals()` would be wrong
    * result of `global x; print(x)` would be wrong if `x` appears in both global and local namespace (if local is preferred, same issue if global is preferred).
    * impossible to do writeback correctly
* use the new method only?
    * the new method will give different exception on some cases (`UnboundLocalError` vs `NameError`)
    * the new method does not work well with `global x; print(x)` as `x` would be defined as both `global` and `nonlocal` (it will fallback to the original method and work)

This is not the perfect solution, but it's pretty close. It passed all the existing tests and fixed all the cases in the issues mentioned above. I have not came up with an example that does not work with this solution (I would guess there will be some dark corners).

Yes, I know this is complicated, and it kind of depends on the implementation of CPython, but I still believe we should do this, because this issue has been repeatedly raised by users of `pdb` for probably > 10 years.

* Issue: gh-83151

### CODE REVIEW

**Comment by gvanrossum** [comment]
Great work! I'm not sure I have the time (or even, at this point, the context) to review this. Maybe someone else can do this more justice?

**Comment by gaogaotiantian** [comment]
> Great work! I'm not sure I have the time (or even, at this point, the context) to review this. Maybe someone else can do this more justice?

Oh I invited you because you reopened https://github.com/python/cpython/issues/65360#issuecomment-1379677554 and this fix involves some black magic of Python. But no worries I'm sure I can find great reviewers for feedbacks of the PR (a couple invited already)!

File: Lib/pdb.py
Comment (brandtbucher) [inline Lib/pdb.py]:
```suggestion
        try:
            compile(source, "<string>", "eval")
        except SyntaxError:
            pass
        else:
            source = "__pdb_eval_result__ = " + source
            locals["__pdb_eval_result__"] = None
```

Comment (brandtbucher) [reply Lib/pdb.py]: Also, why it `__pdb_eval_result__` being set to `None`? You handle the nonexistent case when you pop it later.

File: Lib/pdb.py
Comment (brandtbucher) [inline Lib/pdb.py]:
```suggestion
        source = ("try:\n" +
                  textwrap.indent(source, "  ") + "\n" +
                  "finally:\n" +
                  "  __pdb_write_back__ = locals()")
```

File: Lib/pdb.py
Comment (brandtbucher) [inline Lib/pdb.py]: You can use `update` instead of the loop. Combined with the suggestion below:

```suggestion
            locals.update(write_back)
```

File: Lib/pdb.py
Comment (brandtbucher) [inline Lib/pdb.py]: I'd get rid of the huge outer try/except, and just keep things nice and tight here:

```suggestion
            try:
                exec(code, globals, locals, closure=cells)
            finally:
                write_back = locals.pop("__pdb_write_back__", {})
                eval_result = locals.pop("__pdb_eval_result__", None)
                
```

File: Lib/pdb.py
Comment (brandtbucher) [inline Lib/pdb.py]: I think this line is a little *too* clever. Thankfully, after my suggestion above, we have `eval_result` in a local now.

File: Lib/pdb.py
Comment (brandtbucher) [inline Lib/pdb.py]: I think we can do this without rooting around in `co_consts`:

```suggestion
            # Build a closure source code with freevars from locals like:
            # def __pdb_nonlocal_scope__():
            #   var = None
            #   def __pdb_local_scope__():  # This is the code object we want to execute
            #     nonlocal var
            #     <source>
            #   return __pdb_local_scope__.__code__
            source_with_closure = ("def __pdb_nonlocal_scope__():\n" +
                                   "\n".join(f"  {var} = None" for var in local_vars) + "\n" +
                                   "  def __pdb_local_scope__():\n" +
                                   "\n".join(f"    nonlocal {var}" for var in local_vars) + "\n" +
                                   textwrap.indent(source, "    ") +
                                   "\n  return __pdb_local_scope__.__code__"
                                   )

            # Compile the instrumented source code, and get the code object of __pdb_scope()
            # This simulates executing the original source code with locals as cellvars:
            code = exec(source_with_closure, globals, locals)
            code = locals.pop("__pdb_nonlocal_scope__")()
```

**Review by brandtbucher — COMMENTED** [review]
Been playing around with this locally for a while, and it seems like it works really well! Nice job.

Here's a first round of review. This is a tricky task, but I think the current solution may be overcomplicating it in places:

File: Misc/NEWS.d/next/Library/2023-10-20-03-50-17.gh-issue-83151.bcsD40.rst
Comment (brandtbucher) [inline Misc/NEWS.d/next/Library/2023-10-20-03-50-17.gh-issue-83151.bcsD40.rst]: I'd flesh this out a bit. This is a really nice fix!

File: Lib/pdb.py
Comment (brandtbucher) [inline Lib/pdb.py]: Remind me why this was moved?

File: Lib/pdb.py
Comment (gaogaotiantian) [reply Lib/pdb.py]: Is this correct? The code you got is a function I think? And you are supposed to get the return value of the function?

Comment (brandtbucher) [reply Lib/pdb.py]: I modified the outer function to return the inner function’s code. Then we fish the outer function out of the locals, and call it to get the code.

File: Lib/pdb.py
Comment (gaogaotiantian) [reply Lib/pdb.py]: `buffer` is needed below and if the `if` statement is `False` then `buffer` will never be initialized. You need to expand the full code to not be tricked :)

**Comment by gaogaotiantian** [comment]
I just realized this might have some conflict with PEP 667 which will probably be merged today. I will rework the code. Also taking the notes from your reviews. I will also hopefully finish this tomorrow.

File: Lib/pdb.py
Comment (gaogaotiantian) [reply Lib/pdb.py]: Where did you call the outer function?

Comment (brandtbucher) [reply Lib/pdb.py]: `code = locals.pop("__pdb_nonlocal_scope__")()`. Note the trailing parens.

I guess you could also call it in the source, then save the result in a local. This just seemed cleaner.

**Comment by gaogaotiantian** [comment]
I had to make quite a bit of changes to make this work with PEP 667 - `locals()` is not write-through anymore for function scope. Neither can it do `pop()`. So a couple of things were changed:

1. `locals` is copied first and the writeback only happens at the end of the function.
2. Whether there is nested function is checked first, if not, use the normal `exec`. This actually solves plenty of interesting issues like `locals()` (which will print the extra helper variables without the protection).
3. Some of the reviews are applied.
4. We can't use `locals()` to pass around data anymore, so everything we need is now stored in a dict `__pdb_eval__`.

This relies on the fix from #118583 so please do not merge it yet. I applied the change from that issue just to make the tests pass. I'll merge it in this PR when that one is merged in main.

File: Lib/pdb.py
Comment (brandtbucher) [inline Lib/pdb.py]: Nice simplification to check for code objects here.

I find this a little clearer than `for`/`break`/`else`:

```suggestion
        if any(isinstance(const, CodeType) for const in code.co_consts):
            return False
```

File: Misc/NEWS.d/next/Library/2023-10-20-03-50-17.gh-issue-83151.bcsD40.rst
Comment (gaogaotiantian) [reply Misc/NEWS.d/next/Library/2023-10-20-03-50-17.gh-issue-83151.bcsD40.rst]: Okay I tried to explain it better, but I'm not sure if I did it well. Do you think we need a whatsnews entry for it? I think the news entry itself does not need too much polishing as .. well, not a lot of people are reading it.

File: Lib/pdb.py
Comment (brandtbucher) [inline Lib/pdb.py]: Don't need the `.keys()` here. But also, I don't think you need this at all... you can just loop over `locals_copy` wherever `local_vars` is used below:

```suggestion
```

File: Lib/pdb.py
Comment (brandtbucher) [inline Lib/pdb.py]: This would work, right?

```suggestion
                  "  __pdb_eval__['write_back'] = locals()")
```

File: Lib/pdb.py
Comment (brandtbucher) [inline Lib/pdb.py]:
```suggestion
        exec(source_with_closure, {}, locals_copy)
```

File: Lib/pdb.py
Comment (brandtbucher) [inline Lib/pdb.py:697]: If this raises, we should just let it propagate, right? Otherwise we fall back on `exec`'ing it normally which could change the exception? Or are there cases where this could fail and the correct evaluation isn't in a closure?

```suggestion
        exec(code, globals, locals_copy, closure=cells)
```

File: Lib/pdb.py
Comment (brandtbucher) [inline Lib/pdb.py]: No need to pop:

```suggestion
        pdb_eval = locals_copy["__pdb_eval__]
```

File: Lib/pdb.py
Comment (brandtbucher) [inline Lib/pdb.py:703]: My brain hurts... Does the write-back dict really contain itself? Trippy.

File: Lib/pdb.py
Comment (brandtbucher) [reply Lib/pdb.py:697]: And if this is true, it's probably cleaner to separate this function into two: one for deciding if we need to run in a closure, and another for doing it.

**Review by brandtbucher — COMMENTED** [review]
I like how this is starting to look. Some more thoughts:

File: Lib/pdb.py
Comment (gaogaotiantian) [reply Lib/pdb.py:697]: This is intentional - any exception in the new method, we fall back to the old pure `exec`. There are two reasons behind it:
1. To keep maximum backwards compatibility, making sure this is an incremental improvement.
2. Because of the extra code secretly added, some of the user code may conflict with it and they will have no idea when they saw the exception. For example, if they do somethign like `global x`, which conflicts with `nonlocal x`.

File: Lib/pdb.py
Comment (gaogaotiantian) [reply Lib/pdb.py:703]: The write-back dict is updated with `locals()` which contains the `__pdb_eval__` variable. Yes, this part is a bit twisted.

File: Lib/pdb.py
Comment (gaogaotiantian) [reply Lib/pdb.py]: After PEP 667 I think so. I'm hesitated to use `locals()` directly because it is very complicated before PEP 667, but now it should only be a snapshot.

File: Lib/pdb.py
Comment (gaogaotiantian) [reply Lib/pdb.py]: You are right, this variable is not needed anymore.

**Comment by gaogaotiantian** [comment]
We need to put the other `exec` in the try .. except block as well. The newly added test case is for that. Basically if we have a global variable `x` and we do

```
global x; (lambda: x)()
```

We should be able to get the value. But compiling this with our instrumentation would be:

```
nonlocal x
global x; (lambda: x)()
```

and returns an error that's confusing to users - `*** SyntaxError: name 'g' is nonlocal and global`

That's why we need to fall back to normal.

File: Lib/pdb.py
Comment (brandtbucher) [reply Lib/pdb.py:697]: > 2\. For example, if they do somethign like `global x`, which conflicts with `nonlocal x`.

But that would fail earlier, when `exec`'ing `source_with_closure`, right? Are there other cases where this can raise incorrectly?

What I'm concerned about is if the code the user's running has a bug, and instead of raising the "correct" error here, we swallow it, and give them an "incorrect" error (or worse, silently succeed) by falling back on a non-closure `exec`.

If you're confident that can't happen, though, then this is fine.

File: Lib/pdb.py
Comment (brandtbucher) [inline Lib/pdb.py:651]: Do we need to initialize these, or are there really cases where the `exec` may succeed without these being set?
```suggestion
        locals_copy["__pdb_eval__"] = {}
```

File: Lib/pdb.py
Comment (brandtbucher) [inline Lib/pdb.py]: Does this need to be `locals_copy`?

```suggestion
        ns = {}
        try:
            exec(source_with_closure, {}, ns)
        except Exception:
            return False
        code = ns["__pdb_outer"]()
```

File: Lib/pdb.py
Comment (brandtbucher) [inline Lib/pdb.py]:
```suggestion
        # def __pdb_outer():
```

File: Lib/pdb.py
Comment (brandtbucher) [reply Lib/pdb.py:651]: In fact, do we need this nested dictionary at all? Or can we just assign to `__pdb_eval__` and `__pdb_write_back__` in the generated code directly, and pop those out of the locals? I think that might make it a bit easier to follow, since it removes a level of nesting.

Comment (brandtbucher) [reply Lib/pdb.py:651]: (In fact I think that's what an earlier version of this PR did.)

Comment (gaogaotiantian) [reply Lib/pdb.py:651]: locals will not be updated. locals is only used to initialized the fast variables in the frame and they were never converted back.

File: Lib/pdb.py
Comment (gaogaotiantian) [reply Lib/pdb.py:697]: I can't think of good examples here, but in my mind, the users always get the "wrong" error for the code that supposed to have the "correct" error - we didn't change that. Without any concrete example, I'd like to keep things the same in the dark area - do not introduce any new errors to code that's working before, leave the errors that we could report better as it is for now.

An extreme yet not that convincing example would be if the user uses variable like `__pdb_eval__` in their code - that's some artifact we introduced to make this happen, not part of the original code.

If in the future we had a real case where keeping the exception of our new method is helpful, we can change it.

File: Lib/pdb.py
Comment (gaogaotiantian) [reply Lib/pdb.py:651]: Without initializing them, we need to do something like `if pdb_eval.pop("result") is None` after. It's not horrible, but I don't really see the benefit here not initializing them.

File: Lib/pdb.py
Comment (gaogaotiantian) [reply Lib/pdb.py]: It doesn't have to be. `ns` might make the code easier to understand that the original locals is not required. I used it because the namespace is just there. But this may make the code more isolated so I'll apply the change.

File: Lib/pdb.py
Comment (brandtbucher) [reply Lib/pdb.py:697]: Okay, makes sense!

**Review by brandtbucher — APPROVED** [review]
Okay, this seems correct! Thanks for tackling such a longstanding and hairy issue, and for taking the time to walk me through the fix. :)

**Comment by gaogaotiantian** [comment]
I'm glad this finally got fixed :)

**Pull request merged**
